link.go

```diff
@@ -366,27 +400,27 @@ Loop:
 }
 
 // muxFlow sends tr to the session mux.
-func (l *link) muxFlow() error {
-	// copy because sent by pointer below; prevent race
+// l.linkCredit will also be updated to `linkCredit`
+func (l *link) muxFlow(linkCredit uint32, drain bool) error {
 	var (
-		linkCredit    = l.receiver.maxCredit - uint32(l.countUnsettled())
```

Comment:
Where do you keep track of the unsettled messages? That's what is needed for "ModeSecond"
Do you have tests that verifies you don't break the ModeSecond behavior and the flow associated?
The credit should not be released until the disposition has been sent.

Reply:
I concur WRT tests.  We need to get a better/stronger baseline of tests before we start making such changes.

Reply:
@serbrech : This calculation was changed so it happens outside of muxFlow but it's still there and activates in the non-manual credit flow:

https://github.com/Azure/go-amqp/pull/56/files/ab3ddb309afcbe9ca5ab3e39df28c739f57c4746#diff-ff692b0e80c3a46490bdaf004f6628c003a4736c549aa8050a430246e70d6703R318

This was needed so muxFlow doesn't automatically assume it should calculate a new credit value in the manual credit case. The logic is still the same if you don't opt in to manual credit management.

> The credit should not be released until the disposition has been sent.

This is what go-amqp does today, but what I'm changing (for manual credit flows) where the caller is in charge of if/when credits are added.

@jhendrixMSFT : with regards to tests: https://github.com/Azure/go-amqp/pull/56/files/ab3ddb309afcbe9ca5ab3e39df28c739f57c4746#diff-dc23332a233554520a35145c21c597f8c2164c68accd63315250acdc3dbe3a3f

The changes I'm making don't impact the old flow (even my refactors preserve it). It just allocates a new one where if a manual creditor is present it takes over.

Reply:
Also, the important-est thing - this is written such that it does not affect the non-manual credit path at all. Ie, I'm not changing the old behavior, merely carving out a spot for some new behavior. The default is still the older non-manual behavior.